phylo/src/optimisers/branch_length_optimiser.rs

```diff
@@ -1,29 +1,113 @@
 use argmin::core::{CostFunction, Executor, IterState, State};
 use argmin::solver::brent::BrentOpt;
-use log::{debug, info, warn};
+use log::{debug, info};
 
-use crate::substitution_models::{SubstitutionLikelihoodCost, SubstitutionModel};
-use crate::tree::NodeIdx;
+use crate::phylo_info::PhyloInfo;
+use crate::pip_model::{PIPLikelihoodCost, PIPModel};
+use crate::substitution_models::SubstitutionModel;
+use crate::tree::{NodeIdx, Tree};
 use crate::Result;
 
-pub(crate) struct SingleBranchOptimiser<'a> {
-    pub(crate) likelihood_cost: &'a SubstitutionLikelihoodCost<'a, 4>,
-    pub(crate) branch_length: &'a NodeIdx,
-    pub(crate) model: &'a SubstitutionModel<4>,
+pub(crate) struct SingleBranchOptimiser<'a, SubstModel: SubstitutionModel> {
+    pub(crate) model: &'a PIPModel<SubstModel>,
+    pub(crate) phylo_info: &'a PhyloInfo,
+    pub(crate) branch: &'a NodeIdx,
 }
 
-impl CostFunction for SingleBranchOptimiser<'_> {
+impl<'a, SubstModel: SubstitutionModel + Clone> CostFunction
+    for SingleBranchOptimiser<'a, SubstModel>
+where
+    SubstModel::ModelType: Clone,
+    SubstModel::Params: Clone,
+{
     type Param = f64;
     type Output = f64;
 
     fn cost(&self, value: &Self::Param) -> Result<Self::Output> {
-        let mut upd_tree = self.likelihood_cost.info.tree.clone();
-        upd_tree.set_branch_length(*self.branch_length, *value);
-        // self.likelihood_cost.info.tree = upd_tree;
-        Ok(-self.likelihood_cost.compute_log_likelihood(self.model).0)
+        let mut lc = PIPLikelihoodCost {
+            model: self.model,
+            info: self.phylo_info.clone(),
+        };
+        lc.info.tree.set_branch_length(self.branch, *value);
+        Ok(-lc.compute_log_likelihood().0)
     }
 
     fn parallelize(&self) -> bool {
         true
     }
 }
+
+pub struct BranchOptimiser<'a, SubstModel: SubstitutionModel> {
+    pub(crate) epsilon: f64,
+    pub(crate) model: &'a PIPModel<SubstModel>,
+    pub(crate) phylo_info: PhyloInfo,
+}
+
+impl<'a, SubstModel> BranchOptimiser<'a, SubstModel>
+where
+    SubstModel: SubstitutionModel + Clone,
+    SubstModel::ModelType: Clone,
+    SubstModel::Params: Clone,
+{
+    pub fn new(model: &'a PIPModel<SubstModel>, phylo_info: &PhyloInfo) -> Self {
+        BranchOptimiser {
+            epsilon: 1e-3,
+            model,
+            phylo_info: phylo_info.clone(),
+        }
+    }
+
+    pub fn optimise_parameters(&self) -> Result<(u32, Tree, f64, f64)> {
```

Comment:
nit: consider using structures instead of tuples. its to hard to understand what are these u32,Tree,f64,f64. Everything except for Tree is too generic.

Reply:
Alternatively, you can put a comment e.g. `// iterations, tree copy, init, opt` or whatever makes sense

Reply:
Will be fixed in next PR with an optimisation result struct that all optimisers will use.